haystack/preview/components/reader/extractive.py

```diff
@@ -0,0 +1,321 @@
+        stride: int = 128,
+        max_batch_size: Optional[int] = None,
+        answers_per_seq: Optional[int] = None,
+    ) -> None:
+        """
+        Creates an ExtractiveReader
+        :param model: A HuggingFace transformers question answering model.
+            Can either be a path to a folder containing the model files or an identifier for the HF hub
+            Default: `'deepset/roberta-base-squad2-distilled'`
+        :param device: Pytorch device string. Uses GPU by default if available
+        :param top_k: Number of answers to return per query
+            Default: 10
+        :param top_p: Probability mass that should be contained in returned samples (nucleus sampling)
+        :param max_seq_length: Maximum number of tokens.
+            If exceeded by a sequence, the sequence will be split.
+            Default: 128
+        :param stride: Number of tokens that overlap when sequence is split because it exceeds max_seq_length
+            Default: 128
+        :param max_batch_size: Maximum number of samples that are fed through the model at the same time
+        :param answers_per_seq: Number of answer candidates to consider per sequence.
+            This is relevant when a document has been split into multiple sequence due to max_seq_length.
+        """
+        torch_and_transformers_import.check()
+        self.model = str(model)
+        self.model_ = None
+        self.device = device
+        self.max_seq_length = max_seq_length
+        self.top_k = top_k
+        self.top_p = top_p
+        self.stride = stride
+        self.max_batch_size = max_batch_size
+        self.answers_per_seq = answers_per_seq
+
+    def warm_up(self):
+        if self.model_ is None:
+            if torch.cuda.is_available():
+                self.device = self.device or "cuda:0"
+            else:
+                self.device = self.device or "cpu:0"
+            self.model_ = AutoModelForQuestionAnswering.from_pretrained(self.model).to(self.device)
+            self.tokenizer = AutoTokenizer.from_pretrained(self.model)
+
+    def _flatten(
+        self, queries: List[str], documents: List[List[Document]]
+    ) -> Tuple[List[str], List[Document], List[int]]:
+        flattened_queries = [query for documents_, query in zip(documents, queries) for _ in documents_]
+        flattened_documents = [document for documents_ in documents for document in documents_]
+        query_ids = [i for i, documents_ in enumerate(documents) for _ in documents_]
+        return flattened_queries, flattened_documents, query_ids
+
+    def _preprocess(
+        self, queries: List[str], documents: List[Document], max_seq_length: int, query_ids: List[int], stride: int
+    ) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor, List[Encoding], List[int], List[int]]:
+        encodings = self.tokenizer(
+            queries,
+            [document.content for document in documents],
+            padding=True,
+            truncation=True,
```

Comment:
Nice! I wonder if we are getting a speed boost here since calling the HF tokenizer like this will "pad to the longest sequence in the batch (no padding is applied if you only provide a single sequence)" (reference [here](https://huggingface.co/docs/transformers/pad_truncation)). I'm not sure if this was also the case for the FARMReader v1.

Reply:
Yes, you are right. The speed is very similar when using long input texts. I thought wrongly that `FARMReader` was also doing that.